sdk/src/main/java/com/mapbox/maps/MapboxMap.kt

```diff
@@ -167,10 +168,34 @@ class MapboxMap internal constructor(
     onStyleLoaded: Style.OnStyleLoaded? = null
   ) {
     this.style = style
-    styleExtension.images.forEach {
-      it.bindTo(style)
-    }
+    var geojsonSourceCount = 0
+    var hasGeojsonSources = false
     styleExtension.sources.forEach {
+      if (it is GeoJsonSource) {
+        hasGeojsonSources = true
+        geojsonSourceCount++
+        it.applyDataAsync {
+          geojsonSourceCount--
+          it.bindTo(style)
+          if (geojsonSourceCount == 0) {
+            loadAllExceptGeojsonSources(style, styleExtension, onStyleLoaded)
+          }
+        }
+      } else {
+        it.bindTo(style)
+      }
+    }
+    if (!hasGeojsonSources) {
+      loadAllExceptGeojsonSources(style, styleExtension, onStyleLoaded)
+    }
+  }
+
+  private fun loadAllExceptGeojsonSources(
```

Comment:
why don't we set the empty geojsonsource first and then update the geojsonSource data property? we don't need to complicate the process here..

Reply:
We don't want complexity but we will have it in any case 😄 
It's all because that we will still need exact correct moment of time to emit `onStyleLoaded?.onStyleLoaded(style)` when **everything** is loaded. That brings a lot of complexity as we must emit it when all geojson sources are parsed, after that we add everything else in sync manner and could emit `onStyleLoaded?.onStyleLoaded(style)`.

Reply:
Discussed offline, approach was changed.